UserName Sorting

kyeongminkang opened the pull request:
INDEX에서 최신 글이 맨 위로 올라오게 하기 위해 ID(글번호)로 정렬을 해놓은 상태
UserName을 누르면 오름차순으로 정렬되고, 다시 누르면 내림차순으로 정렬

kyeongminkang committed:
UserName 정렬

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -45,11 +46,22 @@ public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuer
             }
 
-            boards = boards.OrderByDescending(s => s.ID);
+            ViewData["NameSortParm"] = sortOrder;
```
이런식 ViewData나 TempData, ViewBag 같은건 잘 사용 안하구, 대신 ViewModel을 만들어줘서 전달하는걸 선호하는 편입니다.
BoardIndexViewModel.cs 이런식으로 만들어놓고, 쓰는 경우가 대부분이에요.
Property는 NameSortParam, Boards(타입은 List<PaginatedList<Board>>)가 되겠네요.

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -45,11 +46,22 @@ public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuer
+            ViewData["NameSortParm"] = sortOrder;
 
-            int pageSize = 5;
+            switch (sortOrder)
```
switch문이랑 string 타입 조합은 위험해서 함부로 사용하지 않습니다.

case에 "name" 이 오는 경우도 있고, "NAME" 이 올수도 있는 등 실수하기 쉬운 구조거든요.

보통 이럴때는 SortOrder 용 enum 을 별도로 만들어주고 그걸 사용합니다. 
[enum 클래스](https://docs.microsoft.com/ko-kr/dotnet/api/system.enum?view=netcore-2.1) 

실제 코드는 어떻게 사용하는지는 Tools 프로젝트 참고하세요.

kyeongminkang replied:
별도로 만든 enum을 string값으로 변환해서 sortOrder 파라미터 값과 비교하는 IF문 형태로 바꿨습니다. 이렇게 진행하는 것이 맞을까요?

devtedlee replied:
아니요. Enum은 enum하고 비교하는게 제일 좋아요. 부득이한 경우 제외하고. 타입을 만드는 이유는 일종의 지켜야할 규격(또는 사양)을 만들기 위함인데, 활용할 수 있으면 해야죠

kyeongminkang replied:
Enum과 enum의 차이를  Enum 클래스형과 안에 요소인 enum으로 이해하는 것이 맞을까요? 파라미터 값을 확인 후, 타입이 일치하면 값을 반환해주는 형태로 바꾸었는데 옳게 수정한 것인지 여쭈어봅니다.

devtedlee commented on MvcBoardApp/MvcBoardApp/Views/Boards/Index.cshtml:
```diff
@@ -21,7 +21,24 @@
         <tr>
             <th>
-                @Html.DisplayNameFor(model => model.First().UserName)
+                @if (Model.SortOrder == null)
```
너무 중복되는 코드가 많아요. sortOrder, CurrentFilter 두개 변수의 값만 바뀌는 로직이네요.

devtedlee reviewed (commented):
좀 더 개선해보시길

kyeongminkang committed:
어제 까먹고 커밋 안 함...
Models/DTO/SortOrder(enum)생성
IndexBoardViewModel생성

kyeongminkang committed:
생각없이 코드 중복된것은 줄이려다가 페이지 넘어갈 때 마다 파라미터 값이 바뀌는 문제가 생김
왜냐하면 파라미터에 바뀌는 값을 전달해 내보냈기 때문!

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/DTO/SortOrder.cs:
```diff
@@ -0,0 +1,13 @@
+    public enum SortOrder
+    {
+        Name,
+        name_desc 
```
이름 규칙 일치시킬것

devtedlee replied:
ESortOrder

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/Entity/Board.cs:
```diff
@@ -25,5 +25,7 @@ public class Board
 
         public int CommentCount { get; set; }
+
+
```
공란 지우세요

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/Entity/Board.cs:
```diff
@@ -25,5 +25,7 @@ public class Board
         public DateTime? WriteDate { get; set; }
 
         public int CommentCount { get; set; }
+
```
필요없는 공란이나 줄은 지우시길

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -45,11 +47,28 @@ public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuer
-            int pageSize = 5;
+            IndexBoardViewModel indexBoardViewModel = new IndexBoardViewModel()
+            {
+                
```
공란

kyeongminkang replied:
공란을 제거하면서, ViewModel이 들어가는 윗부분은 공란을 두었습니다. 모든 공란을 제거해야하는걸까요? Tools를 보는데, 명확한 공란의 기준이 이해가가지 않아 질문드립니다.

devtedlee replied:
쓸데없는 공란을 지우는거에요. 로직상 분리가 필요하거나 하면 한줄 띄는건 그렇게 안보구요. 이유없이 두 줄 띄거나 그런것

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -45,11 +47,28 @@ public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuer
+                
+                SortOrder = sortOrder,
+                Boards = await PaginatedList<Board>.CreateAsync(boards.AsNoTracking(), pageNumber ?? 1, PAGE_SIZE, sortOrder)
+                
```
공란

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -8,6 +8,7 @@
 using MvcBoardApp.Models;
 using MvcBoardApp.Models.ViewModels;
+using MvcBoardApp.Models.DTO;
 
```
공란

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/ViewModels/IndexBoardViewModel.cs:
```diff
@@ -0,0 +1,19 @@
+        public PaginatedList<Board> Boards { get; set; }
+       
+        public string SortOrder { get; set; }
+        
```
공란

devtedlee commented on MvcBoardApp/MvcBoardApp/Views/Boards/Index.cshtml:
```diff
@@ -21,32 +22,43 @@
         <tr>
             <th>
-                @Html.DisplayNameFor(model => model.First().UserName)
+                @if (Model.SortOrder == null || Model.SortOrder == "name_desc")
```
ESortOrder 만들어놓은거 사용해주세요

kyeongminkang replied:
타입을 설정하지 않아도 enum에서 타입을 읽어올 수 있다고 하셔서 
말씀하신대로 EsorOrder.Name로 읽어오려고 했습니다.
그랬더니 "EsortOrder.Name 멤버는 인스턴스 참조를 사용하여 액세스할 수 없습니다.
대신 형식 이름을 사용하여 한정하세요." 라는 오류문구가 떠서
enum을 class로 감싸준 후 static 한정자로 받아오니 가능해졌습니다.

또한, 어제처럼 매개변수에 직접 string을 입력하지 않고, enum객체로 불러오긴 했는데
enum은 기본형식이 int라고 하는데 어제 말씀하신 string으로 변환없이 쓰는 부분은 이해가 되지 않아 질문 드립니다.
어제 설명해주신 부분인데 이해를 잘 못한것 같습니다. 시간도 오래걸리고 재질문하여 죄송합니다.

devtedlee replied:
view.cshtml 상단에 using으로 enum 클래스가 있는 네임스페이스를 선언해주시면 사용할 수 있죠.

enum을 선언하면 string과 int가 키값으로 연결되듯이 (자료구조 Dictionary처럼) 정의됩니다. 근데 막상 url에서 enum값을 넘기거나 사용하면 string 으로 들어간다는 의미였어요. 해보고 물어보시는건가요?

devtedlee replied:
지금 enum의 용도에서 class로 감싸서 쓸거면 enum을 왜 쓰는지에 대해 설명한 부분이랑 충돌하죠? 어제 설명을 너무 제대로 안들으셨네요. 제가 잘못설명해서 그럴수도 있으니. 이 부분은 다른 방법을 생각해보는것도 좋을거같아요. 제가 똑같은 내용을 말로도 하고 글로도 하고 말로도 다시 하고, 이런 방식은 별로라고 생각하거든요.

다 글로 정리해서 설명해드리고, 경민님이 글을 보다가 질문 있으면 글을 남기는 편이 더 나을까요?

devtedlee replied:
아니면 그냥 직접 검색해보시는 것도 좋습니다.

kyeongminkang replied:
이번에 제가 이해를 제대로 못했음에도 그 자리에서 바로 질문을 못 드려서 일을 번거롭게 해드린 점 죄송합니다.  안 그래도 저로 인해 시간을 많이 할애해 주시는데 최대한 효율적인 방법을 조금 더 고민해보겠습니다.

kyeongminkang committed:
Enum 비교! 맞는 지 모르겠지만...

devtedlee reviewed (commented):
개선

kyeongminkang committed:
Enum으로 바꿔줄 것!!! 나갈 시간이라서 일단 급하게 커밋!!

kyeongminkang committed:
공란 제거
저번에 하던 정렬을 구현
-> SortOrder 파라미터를 사용하는 데 있어 오타 및 실수를 방지하기 위해 Enum을 사용
-> Enum을 string값으로 변환해서 파라미터와 비교
-> Enum과 비교하는 것이 좋다고 하셔서 시도 중
Enum을 사용하는 이유는 타입을 따로 설정하지 않아도 타입을 읽어오기 때문!
하지만 기본이 int라고 하는데 이것이 어떻게 string으로 읽어오는 지 모르겠음.....
또한, EsorOrder.Name이 가능하다하셔서 시도를 했는데, 인스턴스 참조가 안 된다고 해서 Class로 감싸주고, Static으로 받아왔는데 이것이 맞는 지도 의문...

kyeongminkang committed:
공란 제거 추가

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -8,25 +8,24 @@
 using Microsoft.EntityFrameworkCore;
 using MvcBoardApp.Models;
 using MvcBoardApp.Models.ViewModels;
+using static MvcBoardApp.Models.DTO.ESrotOrder;
```
using MvcBoardApp.Models.DTO; 이것만 해도 가져와서 쓸 수 있을텐데요.

오타 ESrotOrder -> ESortOrder

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -35,35 +34,41 @@ public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuer
+            }
+            indexBoardViewModel.EnumSortOrder = eSortOrder;
+            indexBoardViewModel.Boards = await PaginatedList<Board>.CreateAsync(boards.AsNoTracking(), pageNumber ?? 1, PAGE_SIZE, eSortOrder);
+            return View(indexBoardViewModel);
```
여기는 띄는게 더 코드 보기 좋을거같네요. View로 넘겨주는 주요한 부분이고 로직 맥락이 이어지지 않으니까요.

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -138,7 +135,6 @@ public async Task<IActionResult> Edit([FromRoute]int? ID, [FromRoute]int pageNum
                 PageIndex = pageNumber
             };
-
             return View(editBoardViewModel);
```
줄을 너무 일괄적으로 다 없앴네요. 다시 고치시길.

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -176,10 +169,8 @@ public async Task<IActionResult> Edit([FromRoute]int ID, [FromForm]EditBoardView
                     }
                 }
-
                 return RedirectToAction("Index", "Boards", new { pageNumber });
```
"Index" 이런식으로 쓰는것보다 `nameof(Index)` 이런식으로 쓰는게 더 string 실수를 줄일 수 있어요.

kyeongminkang replied:
BoardController에 있는 것은 수정을 했습니다. CommentController에서 BoardController를 사용? 할 때 인터넷으로 검색 시 대부분 현재 방법을 사용하고, 아니면 nameof를 사용하는 경우 nameof(BoardController.Index), nameof(BoardController).replace("Controller","") 이런 형태로 뒤에 강제로 컨트롤러 글자를 빼서 라우트를 맞추는 작업을 하던데 제가 방법을 찾지 못한 것인지, 혹시 방법이 없다면 기존처럼 사용하는 게 나은지, nameof를 사용하는 것이 나은 지 궁금합니다.(기초적인 것이지만 둘 다 string을 직접 타이핑하는 형태라 여쭈어봅니다.)

devtedlee replied:
컨트롤러 바꿀때는 "Board"로 써줘야합니다. ,,nameof 자체가 이름을 가져와서 쓰는거다보니 안될때는 그냥 string 써야되요. 리플레이스 쓰는 방법도 있긴하겠지만 좋아보이진 않네요

kyeongminkang committed:
nameof로 변경 중에 코멘트를 수정하면 원본 글 상세페이지에서 이전페이지로 넘어가는 데 오류가 있었음.
수정 시에 페이지번호가 넘어가지 않아서 그랬음.
수정했음.
일단 커밋

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -8,25 +8,24 @@
 
         [HttpGet]
-        public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuery]string sortOrder, [FromQuery]string currentFilter, [FromQuery]int? pageNumber)
+        public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuery]ESortOrder eSortOrder, [FromQuery]string currentFilter, [FromQuery]int? pageNumber)
```
eSortOrder 말고 sortOrder로 네이밍 하세요. 타입에서만 e 붙이는거에요

devtedlee commented on MvcBoardApp/MvcBoardApp/Controllers/BoardsController.cs:
```diff
@@ -8,25 +8,24 @@
-        public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuery]string sortOrder, [FromQuery]string currentFilter, [FromQuery]int? pageNumber)
+        public async Task<IActionResult> Index([FromQuery]string searchString, [FromQuery]ESortOrder eSortOrder, [FromQuery]string currentFilter, [FromQuery]int? pageNumber)
         {
-            ViewData["CuurentSort"] = sortOrder;
```
Current

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/DTO/ESortOrder.cs:
```diff
@@ -0,0 +1,16 @@
+        {
+            Basic,
+            Name,
+            Name_desc
```
NameDesc

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/DTO/ESortOrder.cs:
```diff
@@ -0,0 +1,16 @@
+    
+        public enum ESortOrder
+        {
+            Basic,
```
이거 basic이 왜 필요하죠?

kyeongminkang replied:
정렬을 하지 않았을 경우, ID를 기준으로 정렬을 하는데, 원래 viewdata[CurrentSort]를 따로 두다가 viewdata를 없애고 enum에 합치?면서 Basic으로 네이밍을 했습니다. 현재는 IDDesc로 변경하였습니다.

devtedlee commented on MvcBoardApp/MvcBoardApp/Views/Boards/Index.cshtml:
```diff
@@ -21,32 +23,43 @@
         <tr>
             <th>
-                @Html.DisplayNameFor(model => model.First().UserName)
+                @if (Model.EnumSortOrder == ESortOrder.Name || Model.EnumSortOrder == ESortOrder.Basic)
```
EnumSortOrder 말고 SortOrder...

devtedlee commented on MvcBoardApp/MvcBoardApp/Models/ViewModels/IndexBoardViewModel.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public PaginatedList<Board> Boards { get; set; }
+      
+        public ESortOrder EnumSortOrder { get; set; }        
```
EnumSortOrder 》》 SortOrder

devtedlee reviewed (commented):
네이밍에 좀 더 고민을 하면서 하시길

kyeongminkang committed:
공란 및 개행 수정
변수명 변경

devtedlee reviewed (approved):
이정도로 마무리 하죠. 이제 다음 스텝으로 넘어가야 될 시기가 됐습니다. 수고하셨어요.
